Added capability to relaunch an execution

pmahindrakar-oss opened the pull request:
Signed-off-by: Prafulla Mahindrakar <prafulla.mahindrakar@gmail.com>

# TL;DR
Relaunches an execution from create command.

```
flytectl create execution --relaunch xb5317xbty  -p flytectldemo -d development
Warn: No metricsProvider set for the workqueue
INFO[0000] Using config file: [config.yaml]             
INFO[0000] Config section [root] updated. No update handler registered.  src="viper.go:318"
INFO[0000] Config section [logger] updated. Firing updated event.  src="viper.go:320"
{"json":{"src":"viper.go:320"},"level":"info","msg":"Config section [admin] updated. Firing updated event.","ts":"2021-04-14T18:48:11+05:30"}
{"json":{"src":"config.go:52"},"level":"warning","msg":"Admin client config has authentication ON with server ","ts":"2021-04-14T18:48:11+05:30"}
{"json":{"src":"viper.go:318"},"level":"info","msg":"Config section [storage] updated. No update handler registered.","ts":"2021-04-14T18:48:11+05:30"}
{"json":{"src":"viper.go:318"},"level":"info","msg":"Config section [adminutils] updated. No update handler registered.","ts":"2021-04-14T18:48:11+05:30"}
{"json":{"src":"client.go:30"},"level":"info","msg":"Initialized Admin client","ts":"2021-04-14T18:48:11+05:30"}
{"json":{"src":"execution.go:122"},"level":"info","msg":"execution config xb5317xbty of type 2","ts":"2021-04-14T18:48:11+05:30"}
{"json":{"src":"execution_util.go:81"},"level":"debug","msg":"fetched execution xb5317xbty with project flytectldemo and domain development","ts":"2021-04-14T18:48:11+05:30"}
{"json":{"src":"execution_util.go:88"},"level":"debug","msg":"fetched launch plan with name core.advanced.run_merge_sort.merge_sort project flytectldemo domain development and version v1","ts":"2021-04-14T18:48:11+05:30"}
execution identifier project:"flytectldemo" domain:"development" name:"fffd97c78a818411cb85"
```

## Type
 - [ ] Bug Fix
 - [X] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [X] Unit tests added
 - [X] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/915

## Follow-up issue
_NA_
OR
_https://github.com/lyft/flyte/issues/<number>_

katrogan commented on cmd/create/execution.go:
```diff
@@ -110,15 +126,23 @@ func createExecutionCommand(ctx context.Context, args []string, cmdCtx cmdCore.C
 	if execParams, err = readConfigAndValidate(config.GetConfig().Project, config.GetConfig().Domain); err != nil {
 		return err
 	}
+	logger.Infof(ctx, "execution config %v of type %v", execParams.name, execParams.execType)
```
remove?

pmahindrakar-oss replied:
Done

katrogan commented on cmd/create/execution_test.go:
```diff
@@ -147,20 +180,65 @@ func TestCreateLaunchPlanExecutionFunc(t *testing.T) {
-	mockClient.OnCreateExecutionMatch(ctx, mock.Anything).Return(executionCreateResponseTask, nil)
-	executionConfig.ExecFile = testDataFolder + "task_execution_spec.yaml"
+	literalMap := &core.LiteralMap{
+		Literals: nil,
```
nit: do you need to bother initializing to nil?

pmahindrakar-oss replied:
Yeah not required. I had some params earlier but not required for the test and hence removed it

katrogan commented on cmd/create/execution_test.go:
```diff
@@ -147,20 +180,65 @@ func TestCreateLaunchPlanExecutionFunc(t *testing.T) {
+	mockFetcher.OnFetchExecutionMatch(mock.Anything, mock.Anything, mock.Anything, mock.Anything, mock.Anything).Return(exec, nil)
+	mockFetcher.OnFetchLPVersionMatch(mock.Anything, mock.Anything, mock.Anything, mock.Anything, mock.Anything, mock.Anything).Return(launchPlan, nil)
+	mockClient.OnCreateExecutionMatch(ctx, mock.Anything).Return(executionCreateResponseLP, nil)
+	executionConfig.Relaunch = "xb5317xbty"
```
is this the execution to be relaunched? how come it doesn't match "ffb31066a0f8b4d52b77",?

pmahindrakar-oss replied:
Since i am using mock.Anything as input hence it didnt require a match . Have added an exact match now

katrogan commented on cmd/create/execution_util.go:
```diff
@@ -70,6 +70,29 @@ func createExecutionRequestForTask(ctx context.Context, taskName string, project
 	return createExecutionRequest(ID, inputs, authRole), nil
 }
 
+func createExecutionRequestForRelaunch(ctx context.Context, executionName string, project string, domain string, cmdCtx cmdCore.CommandContext) (*admin.ExecutionCreateRequest, error) {
```
can we use the RelaunchExecution api here? (https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/admin/execution.proto#L37) this way you don't have to fetch the execution and the lp

pmahindrakar-oss replied:
Done

kumare3 commented on cmd/create/execution.go:
```diff
@@ -72,6 +72,12 @@ The root project and domain flags of -p and -d should point to task/launch plans
+
+::
+
+ flytectl create execution --relaunch ffb31066a0f8b4d52b77 -p flytectldemo -d development
```
can you also supply a new name as the argument - Guess you can right as you do in regular launch execution?

pmahindrakar-oss replied:
Will be doing in a followup PR.

kumare3 commented on cmd/get/get.go:
```diff
@@ -17,6 +18,17 @@ Example get projects.
+	return FetcherImpl{}
+}
+
+type FetcherImpl struct {
```
is this only for testing? the cmdCtx being passed seems odd. Why not make this part of the command ?

pmahindrakar-oss replied:
This was supposed to be used by create module when it needs to use the get functionalities.
Hence defined this interface for it.
The relaunch execution earlier requires the get for lp and exec but now we can do just directly relaunch.

But plain exec still relies on get functionalities and i was planning to add more things to this interface so that create can use this

pmahindrakar-oss replied:
Will take the feedback in the followup PR.

kumare3 reviewed (dismissed):
lgtm

katrogan commented on cmd/create/execution_util.go:
```diff
@@ -70,6 +68,21 @@ func createExecutionRequestForTask(ctx context.Context, taskName string, project
 }
 
+func relaunchExecution(ctx context.Context, executionName string, project string, domain string, cmdCtx cmdCore.CommandContext) error {
+	relaunchedExec, _err := cmdCtx.AdminClient().RelaunchExecution(ctx, &admin.ExecutionRelaunchRequest{
```
meganit :), `_err` to `err`?

pmahindrakar-oss replied:
Done

katrogan commented on cmd/get/interfaces/execution_interface.go:
```diff
@@ -0,0 +1,15 @@
+
+//go:generate mockery -all -case=underscore
+
+type Fetcher interface {
```
What's the purpose of this interface? Can you add comments? Also, why is the file called execution_interface when it also includes FetchLPVersion?

pmahindrakar-oss replied:
Added the comments for it. 
Its used for exposing the create capabilities to other command modules.
eg : create exeuction which requires to fetch the launcplan for its parameters details can use this interface to fetch the data.

katrogan reviewed (dismissed):
looks good, just a comment & suggestion

pmahindrakar-oss commented:
@katrogan / @kumare3 can you re-approve this PR . Unfortunately when i push commits after addressing the review comments, the github policy on this repo dismisses the old approvals .

https://github.com/MakeContributions/DSA/pull/135
